### PR TITLE
Add function to invert ranges

### PR DESCRIPTION
**Pull request by th4s**
This PR adds a function which allows to invert ranges.

### CODE REVIEW

File: utils/src/invert_range.rs [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+        }
+
+        // Check that ranges are not overlapping
+        if ranges_to_remove
+            .iter()
+            .enumerate()
+            .any(|(l, r)| k != l && r.start < range.end && r.end > range.start)
+        {
+            return Err(RangeError::Overlapping);
+        }
```
Comment (sinui0) [inline utils/src/invert_range.rs]: This is O(n^2). We can probably do better with a sort:

pseudo code:

```rust
ranges_to_remove.sort_by(|r| r.start);
assert!(ranges_to_remove.windows(2).all(|(l, r)| r.start > l.start && r.end > l.end));
```

File: utils/src/invert_range.rs [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+            .expect("Should have found range to invert");
+
+        let original_end = inv.end;
+        inv.end = range.start;
+
+        inverted.push(Range {
+            start: range.end,
+            end: original_end,
+        });
+    }
```
Comment (sinui0) [inline utils/src/invert_range.rs]: With `ranges_to_remove` sorted you can just operate on the tail

```suggestion
    let mut inverted = Vec::with_capacity(ranges_to_remove.len() + 1);
    inverted.push(original_range.clone());
    for remove in ranges_to_remove {
        let tail = inverted.last_mut().unwrap();
        
        tail.end = remove.start;
        
        inverted.push(remove.end..original_range.end);
    }
```

File: utils/src/invert_range.rs [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+
+    // Remove empty ranges
+    inverted.retain(|r| r.start != r.end);
```
Comment (sinui0) [inline utils/src/invert_range.rs:66]: This should be unneeded with above

**Review by sinui0 — CHANGES_REQUESTED** [review]
Perhaps we should move this to `spansy`? I expect this functionality to be exclusively used with spans. It could be implemented as a method on `Span`

**Comment by th4s** [comment]
I also thought about moving this to `spansy`. What kept me from doing it, was that this works with ranges and has nothing to do with spanning. The question is probably if this functionality will be exposed by `tlsn` or by `spansy`.

Considering the performance, I did not pay much attention to it. But makes sense.

File: utils/src/invert_range.rs (hunk 2 above)
Comment (th4s) [reply utils/src/invert_range.rs]: very nice! :+1:

File: utils/src/invert_range.rs (hunk 3 above)
Comment (th4s) [reply utils/src/invert_range.rs:66]: This is still needed if `tail.end == remove.start`.

**Review by sinui0 — APPROVED** [review]

**Comment by themighty1** [comment]
Is the plan to use this approach or the `rangetools` crate?

**Comment by th4s** [comment]
This is superseded by #4

**Pull request closed without merging**